File: day2-langgraph/agent.py
import sys
import os
import asyncio
from typing import TypedDict, Annotated, List, Union

# Add the parent directory to sys.path to allow imports from common
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from common.model import get_model
from langchain_core.messages import HumanMessage, SystemMessage, BaseMessage
from langchain_core.tools import tool
from langchain_community.document_loaders import TextLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_openai import OpenAIEmbeddings
from langchain_community.vectorstores import FAISS
from langgraph.graph import StateGraph, START, END
from langgraph.prebuilt import ToolNode, tools_condition
from langgraph.graph.message import add_messages

# Define State
class AgentState(TypedDict):
    messages: Annotated[List[BaseMessage], add_messages]

# --- Setup Vector Store (Same as Day 1) ---
async def setup_vector_store():
    current_dir = os.path.dirname(os.path.abspath(__file__))
    file_path = os.path.join(current_dir, "../day1-foundations/info.txt")
    loader = TextLoader(file_path)
    docs = await asyncio.to_thread(loader.load)
    
    splitter = RecursiveCharacterTextSplitter(chunk_size=200, chunk_overlap=20)
    split_docs = splitter.split_documents(docs)
    
    embeddings = OpenAIEmbeddings()
    return await asyncio.to_thread(FAISS.from_documents, split_docs, embeddings)

async def run():
    vector_store = await setup_vector_store()
    retriever = vector_store.as_retriever()

    # --- Define Tools ---
    @tool
    async def lookup_policy(query: str) -> str:
        """Ask this tool questions about LangGraph."""
        docs = await retriever.ainvoke(query)
        return "\n\n".join([d.page_content for d in docs])

    tools = [lookup_policy]
    tool_node = ToolNode(tools)

    # --- Define Graph ---
    model = get_model().bind_tools(tools)

    # Node: Agent (Calls Model)
    async def call_model(state: AgentState):
        response = await model.ainvoke(state["messages"])
        return {"messages": [response]}

    workflow = StateGraph(AgentState)
    workflow.add_node("agent", call_model)
    workflow.add_node("tools", tool_node)
    
    workflow.add_edge(START, "agent")
    workflow.add_conditional_edges("agent", tools_condition)
    workflow.add_edge("tools", "agent")

    app = workflow.compile()

    # --- Run ---
    print("User: What is LangGraph inspired by?")
    result = await app.ainvoke({
        "messages": [HumanMessage(content="What is LangGraph inspired by?")]
    })

    print("\nFinal Response:")
    print(result["messages"][-1].content)

if __name__ == "__main__":
    asyncio.run(run())

File: day2-langgraph/agent.js
import { getModel } from "../common/model.js";
import { StateGraph, MessagesAnnotation } from "@langchain/langgraph";
import { ToolNode } from "@langchain/langgraph/prebuilt";
import { tool } from "@langchain/core/tools";
import { z } from "zod";
import { HumanMessage } from "@langchain/core/messages";
import { MemoryVectorStore } from "langchain/vectorstores/memory";
import { OpenAIEmbeddings } from "@langchain/openai";
import { TextLoader } from "langchain/document_loaders/fs/text";
import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// --- Setup Vector Store (Same as Day 1) ---
const setupVectorStore = async () => {
    const loader = new TextLoader(path.join(__dirname, "../day1-foundations/info.txt"));
    const docs = await loader.load();
    const splitter = new RecursiveCharacterTextSplitter({ chunkSize: 200, chunkOverlap: 20 });
    const splitDocs = await splitter.splitDocuments(docs);
    return await MemoryVectorStore.fromDocuments(splitDocs, new OpenAIEmbeddings());
};

const run = async () => {
    const vectorStore = await setupVectorStore();
    const retriever = vectorStore.asRetriever();

    // --- Define Tools ---
    const lookupPolicy = tool(
        async ({ query }) => {
            const docs = await retriever.invoke(query);
            const content = docs.map((d) => d.pageContent).join("\n\n");
            // console.log("\n[Tool] Retrieved Documents:");
            // console.log(content);
            return content;
        },
        {
            name: "lookup_policy",
            description: "Ask this tool questions about LangGraph.",
            schema: z.object({
                query: z.string().describe("The search query"),
            }),
        }
    );

    const tools = [lookupPolicy];
    const toolNode = new ToolNode(tools);

    // --- Define Graph ---
    const model = getModel().bindTools(tools);

    // Node: Agent (Calls Model)
    const callModel = async (state) => {
        const response = await model.invoke(state.messages);
        return { messages: [response] };
    };

    // Edge: Should Continue?
    const shouldContinue = (state) => {
        const lastMessage = state.messages[state.messages.length - 1];
        if (lastMessage.tool_calls?.length) {
            return "tools";
        }
        return "__end__";
    };

    const workflow = new StateGraph(MessagesAnnotation)
        .addNode("agent", callModel)
        .addNode("tools", toolNode)
        .addEdge("__start__", "agent")
        .addConditionalEdges("agent", shouldContinue)
        .addEdge("tools", "agent");

    const app = workflow.compile();

    // --- Run ---
    console.log("User: What is LangGraph inspired by? What is the latest version available?");
    const result = await app.invoke({
        messages: [new HumanMessage("What is LangGraph inspired by? What is the latest version available?")],
    });

    console.log("\nFinal Response:");
    console.log(result.messages[result.messages.length - 1].content);
};

run().catch(console.error);
